File: tests/pack_unpack_test.go
package tests

import (
	"encoding/json"
	"fmt"
	"testing"
	"time"

	"github.com/bbanez/minibin/src/utils"
	minibin "github.com/bbanez/minibin/tests/dist"
)

func TestPack(t *testing.T) {
	items := minibin.Obj1Arr{
		Items: []*minibin.Obj1{},
	}
	for i := 0; i < 5; i++ {
		item := &minibin.Obj1{
			Str:    fmt.Sprintf("item_%d", i),
			I32:    int32(i * 10),
			I32Arr: []int32{int32(i * 10), int32(i * 20), int32(i * 30)},
			I64:    int64(i * 100),
			I64Arr: []int64{int64(i * 100), int64(i * 200), int64(i * 300)},
			U32:    uint32(i * 1000),
			U32Arr: []uint32{uint32(i * 1000), uint32(i * 2000), uint32(i * 3000)},
			U64:    uint64(i * 10000),
			U64Arr: []uint64{uint64(i * 10000), uint64(i * 20000), uint64(i * 30000)},
			F32:    float32(i) + 0.123,
			F32Arr: []float32{float32(i) + 0.234, float32(i) + 0.345},
			F64:    float64(i) + 0.456789,
			F64Arr: []float64{float64(i) + 0.567890, float64(i) + 0.678901},
			Obj2: minibin.Obj2{
				Key:   fmt.Sprintf("key_%d", i),
				Value: float32(i) * 1.5,
			},
			Obj2Arr: []*minibin.Obj2{
				{
					Key:   fmt.Sprintf("keyarr_%d_1", i),
					Value: float32(i) * 2.5,
				},
				{
					Key:   fmt.Sprintf("keyarr_%d_2", i),
					Value: float32(i) * 3.5,
				},
			},
			Enum1:    minibin.ENUM1_E1,
			Enum1Arr: []minibin.Enum1{minibin.ENUM1_E2, minibin.ENUM1_E3},
		}
		items.Items = append(items.Items, item)
	}
	timeOffset := time.Now().UnixNano()
	s1 := utils.SerializeJson(items)
	fmt.Println("JS time", time.Now().UnixNano()-timeOffset)
	timeOffset = time.Now().UnixNano()
	bytes := items.Pack()
	fmt.Println("MB time", time.Now().UnixNano()-timeOffset)
	fmt.Println("Packed size:", len(bytes), len(s1))
	timeOffset = time.Now().UnixNano()
	e, err := minibin.UnpackObj1Arr(bytes)
	fmt.Println("MB Unpack time", time.Now().UnixNano()-timeOffset)
	if err != nil {
		t.Fatal("Failed to unpack entry:", err)
	}
	timeOffset = time.Now().UnixNano()
	tmp := minibin.Obj1Arr{}
	err = json.Unmarshal([]byte(s1), &tmp)
	fmt.Println("JS Unpack time", time.Now().UnixNano()-timeOffset)
	if err != nil {
		t.Fatal("Failed to unmarshal json:", err)
	}
	s2 := utils.SerializeJson(e)
	if s1 != s2 {
		t.Errorf("Object do not match after pack/unpack")
	}
}

File: tests/ts/pack_unpack.test.ts
import { test } from 'bun:test';
import { Obj1Arr } from '../dist/ts/obj_Obj1Arr';
import { Obj1 } from '../dist/ts/obj_Obj1';
import { Obj2 } from '../dist/ts/obj_Obj2';

// test('pack and unpack small obj', () => {
//     const ob = new Obj2({
//         key: 'key 1',
//         value: 'value 1',
//     });
//     let start = process.hrtime.bigint();
//     const s1 = JSON.stringify(ob, (key, value) => {
//         return typeof value === 'bigint' ? value.toString() : value;
//     });
//     console.log(
//         `JSON pack:`,
//         process.hrtime.bigint() - start,
//         '-> size:',
//         s1.length,
//     );
//     start = process.hrtime.bigint();
//     JSON.parse(s1);
//     console.log('JSON unpack:', process.hrtime.bigint() - start);
//     start = process.hrtime.bigint();
//     let buf = ob.pack();
//     console.log(
//         `MBin pack:`,
//         process.hrtime.bigint() - start,
//         '-> size:',
//         buf.length,
//         buf,
//     );
//     const ob2 = new Obj2({
//         key: 'key 2',
//         value: 'value 2',
//     });
//     start = process.hrtime.bigint();
//     buf = ob2.pack();
//     console.log(
//         `MBin pack 2:`,
//         process.hrtime.bigint() - start,
//         '-> size:',
//         buf.length,
//         buf,
//     );
//     // start = process.hrtime.bigint();
//     // const o2 = Obj2.unpack(ob);
//     // console.log('MBin unpack:', process.hrtime.bigint() - start);
// });

test('pack and unpack data', () => {
    const items = new Obj1Arr({
        items: [],
    });
    for (let i = 0; i < 10000; i++) {
        items.items.push(
            new Obj1({
                str: `Test string ${i}`,
                strArr: [
                    `Item iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii ${i}-1`,
                    `Item ${i}-2`,
                    `Item ${i}-3`,
                ],
                i32: i,
                i32Arr: [i, -i - 1, -i - 200000],
                i64: BigInt(i) * BigInt(100000),
                i64Arr: [BigInt(i), BigInt(-i - 1), BigInt(-i - 200000)],
                u32: i * 2,
                u32Arr: [i * 2, i * 3, i * 4],
                u64: BigInt(i) * BigInt(200000),
                u64Arr: [
                    BigInt(i) * BigInt(2),
                    BigInt(i) * BigInt(3),
                    BigInt(i) * BigInt(4),
                ],
                f32: i + 0.5,
                f32Arr: [i + 0.1, i + 0.2, i + 0.3],
                f64: i + 0.25,
                f64Arr: [i + 0.01, i + 0.02, i + 0.03],
                bool: i % 2 === 0,
                boolArr: [true, false, true],
                enum1: 'E1',
                enum1Arr: ['E1', 'E2', 'E3'],
                obj2: new Obj2({
                    key: `Obj2 key ${i}`,
                    value: i * 10.1,
                }),
                obj2Arr: [
                    new Obj2({
                        key: `Obj2 key ${i}-1`,
                        value: i * 7.2 + 1,
                    }),
                    new Obj2({
                        key: `Obj2 key ${i}-2`,
                        value: i * 8.54003 + 2,
                    }),
                ],
            }),
        );
    }
    for (let i = 0; i < 10; i++) {
        let start = process.hrtime.bigint();
        const s1 = JSON.stringify(items, (_key, value) => {
            return typeof value === 'bigint' ? value.toString() : value;
        });
        console.debug(
            `JSON pack:`,
            (process.hrtime.bigint() - start) / 1000000n,
            'ms -> size:',
            (s1.length / 8 / 1000000).toFixed(2),
        );
        // start = process.hrtime.bigint();
        // const _s1 = JSON.parse(s1);
        // console.debug(`JSON unpack ${i}:`, process.hrtime.bigint() - start);
        start = process.hrtime.bigint();
        const buf = items.pack();
        console.debug(
            `MBin pack ${i}:`,
            (process.hrtime.bigint() - start) / 1000000n,
            'ms -> size:',
            (buf.length / 8 / 1000000).toFixed(2),
            '\n',
        );
        // start = process.hrtime.bigint();
        // Obj1Arr.unpack(buf);
        // console.debug(`MBin unpack ${i}:`, process.hrtime.bigint() - start);
    }
});
